### ISSUE TITLE
Default player starts video before ViewTube can act

### ISSUE THREAD

**Issue by haarp** [sebaro/ViewTube#13]
A problem that often appears when ViewTube is used. The default site video player appears, starts loading and playing, and is then replaced by ViewTube.

Would it be possible to load VT earlier in the site loading process with @run-at document-start or @run-at document-body? VT could then add observers to catch the creation of the player element and instantly destroy it.

**Comment by sebaro** [comment]
Yes, it happens because the scripts loads later than the scripts in the page. It does check for the player, not with an observer but with a interval.

I've never tried the "run-at" metadata, the problem is the support for it in other user scripts extensions.
I think GM4 supports only "document-end":
https://wiki.greasespot.net/Metadata_Block#.40run-at
"The only guaranteed working value, in Greasemonkey 4.0."

You can try these in you browser/extension and let me know which works better.
You can also decrease the interval value.

**Comment by haarp** [comment]
Greasemonkey pre-4.0 definitely supports document-start, I have been using it. It looks like they're working on getting old features back into 4.0 too. Other extensions like Violentmonkey already support document-start.

I've tried reducing the timeout for blockVideos() but that didn't seem to solve the problem. Hmm.

**Comment by sebaro** [comment]
You can test the script with "document-start", there's already code that blocks the videos, it uses 500 ms interval for checking. See if that is better than "document-end".

**Comment by haarp** [comment]
I'm trying document-start and I've reduced the 500ms intervals to 50ms for testing. this is with Violentmonkey on Firefox 57.

A console.log() line at the top of the script immediately logs successfully, so the script is executed at the start!

However, it still is very slow at replacing the video on Youtube. It can take over a second. I can't figure out why.

**Comment by sebaro** [comment]
My changes: http://isebaro.com/viewtube/viewtubetest.user.js

**Comment by haarp** [comment]
Thanks. Unfortunately this doesn't solve it either, it still takes over a second after the appearance of the Youtube player to replace it. I wonder if the Javascript engine of the browser is just so bogged down that it delays the script.

**Comment by sebaro** [comment]
I've added MutationObserver which seams to be faster than setInterval but not enough to block the videos to not hear the audio for those with audio at the beginning. Also player replacement as you said still not fast enough.

**Comment by haarp** [comment]
Interestingly, now it sometimes doesn't replace the player at all. I think VT is taking too long to install the MutationObserver. Is it executing a lot of stuff before that? Things that could possibly block the JS thread for a while?

**Comment by sebaro** [comment]
What do you use? It's fine here with Chromium, no user scripts extensions. Videos are blocked and player is replaced but not fast enough.

**Comment by haarp** [comment]
Violentmonkey  2.8.18 on Firefox 57. I notice from the console logs that it doesn't install the MutationObserver when it fails. Weird.

**Comment by sebaro** [comment]
Your issue can't be caused by MutationObserver, even if this doesn't work the setInterval code should.
It could be caused by some changes I've made to the code for player replacement, but I've reverted them, see if it's working.

**Comment by Zero3K** [comment]
I use the script available at https://gist.githubusercontent.com/anonymous/a1df6903bf17215e9bc73fa3dc3ace6d/raw/8dcf3085ee1402713b7370e60bf867cdacb9a868/YouTube_prevent_create.js in order to have only the player that is assigned by ViewTube loaded.

**Comment by sebaro** [comment]
@Zero3K: how does this help?

**Comment by Zero3K** [comment]
It prevents the creation of the default player that Youtube uses.

**Comment by sebaro** [comment]
The user script loads after site's scripts so I don't know how it helps with this issue.
When using it don't you hear the beginning of the videos for some of them? Try live videos.

**Comment by Zero3K** [comment]
I do not hear the beginning of the videos that I have been watching until the alternative player I have chosen has buffered them.

**Comment by sebaro** [comment]
Try with a live/concert video.

**Comment by haarp** [comment]
I tried just blocking Youtube's player script with Ublock.

    www.youtube.com/yts/jsbin/player*/$important,domain=www.youtube.com

This worked very very well in older Firefoxes. Now in FF57, it causes the page to not load completely and vast areas become "unclickable". Makes no sense.

**Comment by haarp** [comment]
Got it working. Turns out Youtube ships different sites depending on which User-Agent your browser sends. So give it some bullshit useragent, then add this to your adblocker:

    www.youtube.com/yts/jsbin/player*/$important,domain=www.youtube.com
    www.youtube.com##div#placeholder-player
    www.youtube.com##div#watch7-sidebar:style(top: 0px !important; margin-top: 0px !important)

Not perfect, but an usable workaround until we can figure out how to make Viewtube prevent the player from loading directly :)

**Comment by sebaro** [comment]
@haarp, open a new issue if is not related to the current one.
Aso if you use other user scripts and/or addons/extensions that breaks the script there's nothing I can do. When the script loads it must find the player element but if some addon blocks the script that creates that element I don't see what I can do about it.

**Comment by haarp** [comment]
No, @sebaro, these were just comments about blocking the YT player in general. It wasn't related to Viewtube. I've removed the misleading comment.

**Comment by sebaro** [comment]
I don't see a perfect way to block the video, "@run-at document-start" helps but not much and parts of the script must be rewritten to work with it.

Your solution may help, I don't use other addons so I can't confirm.

For Firefox I disable all audio and video codec support in "about:config" but this breaks YouTube's new layout.

**Comment by sebaro** [comment]
I've made some changes, the script doesn't reload the page on video change.

Userscript: http://isebaro.com/viewtube/viewtubetest.user.js
WebExtensions: http://isebaro.com/viewtube/viewtube.zip

**Comment by haarp** [comment]
Hey, this seems to work quite well! Doesn't solve the problem during the initial load, but when opening other videos in the same tab, this is great.

Thanks

**Comment by sebaro** [comment]
For the first video is the same as before.
For same page (related) videos the script doesn't have to be reloaded so it should work better.
